design/pipeline-caching.md

```diff
@@ -0,0 +1,157 @@
+# Pipeline Caching
+
+*** Status: Work in Progress ***
+
+We have observed the need for Azure Pipelines to implement a dependency caching mechanism that allows the outputs of steps within a pipeline to be optimized (or even skipped) if a suitable cached version of the outputs of those steps already exists. Caching can be effective provided the cost of determining a cache hit and acquiring the contents of the cache is cheaper than the cost of producing the output again from scratch.
+
+## Proposal
+
+This proposal contains a number of interconnected elements. We have attempted to layer them in this document from the most primitive building blocks up to simplified YAML syntax to make caching "just work" in the majority of cases. Our approach to building a caching mechanism for Azure Pipelines will be to get the fundamental building blocks right and prove we can increase build performance and then evolve the YAML syntax once we know exactly what that syntax will need to describe.
+
+### Restore and Save Cache Tasks
+
+The basic functions of Pipeline Cache will be implemented by two tasks. The ```RestoreCache``` task and the ```SaveCache``` task. Both tasks will specify a collection of files and environment variables which can be used as a key to lookup the cache and identify a hit or a miss. Each task will also take a collection of paths which will be fetched from or stored in the cache associated with that key.
+
+Consider the following Azure Pipelines pipeline defined in YAML:
+
+```yaml
+steps:
+- script: yarn
+- script: yarn build
+```
+
+In this example we are working on a React-based application which was bootstrapped using ```create-react-app```. The output of this command is a directory with about ~28,000 files including the ```node_modules``` directory. On a local developer machine the ```yarn``` command will trigger the installation of any missing packages. Typically this will execute quickly because all packages already exist on the developer machine. However in cloud-based CI environments where build machines are recycled after every build there is no persistent state which means that Yarn must download all packages from scratch.
+
+Applying caching to this pipeline would involve adding a restore cache step and save cache step before and after the ```-script: yarn``` step.
+
+```yaml
+steps:
+- task: RestoreCache@0
+  inputs:
+    keys: |
+      package.json
+      yarn.lock
+    paths: |
+      node_modules
+- script: yarn
+- task: SaveCache@0
+  inputs:
+    keys: |
+      package.json
+      yarn.lock
+    paths: |
+      node_modules
```

Comment:
For restore, I should not have to specify the folder.  I should just say which key I want and the folders are restored to the right location.

Reply:
Yep - I agree. I was thinking in the end we are going to have to have some kind of "layout" metadata stored in the cache entry because we can't even guarantee that all the files in the cache are in the same path. One question I have with that though is what we do around paths that are owned by specific variables.

In scenarios where you are using a private agent, and one agent has a different work directory we'll need to persist the paths as they are specified in the save step - including variables so that they can be substituted at restore time. Can we get the non-expanded paths passed to the task and do the expansion inside the task??

Reply:
@mitchdenny Right now there is no way for you to get the non-expanded paths and for this feature that is going to be an issue.  We should spin up a converasation with @vtbassmatt about what we might be able to do there.

Reply:
Thanks - good to know. I'll follow-up with @vtbassmatt

Reply:
My initial reaction is that for v1, let's not block on this. The feature is *primarily* about the hosted case, where paths don't change.

I need to think some more about how to offer unexpanded variables to tasks. That feature scares me, as our macro expansion story is already complicated enough!

Reply:
OK - I think we'll need to see how we go here. For relative paths in the working directory it should be fine, if we end up having to drop cache into tool directories then that is where the brittleness will creep in. Either way we should get the ball rolling in allowing tasks to get this information.